### PR TITLE
add option in 'Dev Settings' for never use cached bundle file

### LINKED ISSUE #2676: [Workflow][Android] Provide an option to never load JS from cache
Currently RN Android loads JS from a cache if it can't reach the bundle server. This is for teams working on hybrid apps since not everyone is running a packager. But when you're doing RN development you want to always be loading the most recent version of the code, so some option to never load from cache would be good.

### PR DESCRIPTION
**Pull request by qbig**
fix #2676

### CODE REVIEW

File: ReactAndroid/src/main/res/devsupport/xml/preferences.xml [hunk 1]
```diff
@@ -19,6 +19,12 @@
         android:summary="At the end of animations, Toasts and logs to logcat debug information about the FPS during that transition. Currently only supported for transitions (animated navigations)."
         android:defaultValue="false"
         />
+    <CheckBoxPreference
+        android:key="use_cached_bundle"
+        android:title="Use Cached Bundle"
+        android:summary="Use prefetched JS bundle stored in local file."
```
Comment (mkonicek) [inline ReactAndroid/src/main/res/devsupport/xml/preferences.xml]: > stored in local file

"stored locally"

The way it's stored might change, making this message out of date.

File: ReactAndroid/src/main/java/com/facebook/react/devsupport/DevInternalSettings.java [hunk 2]
```diff
@@ -64,6 +65,11 @@ public boolean isJSDevModeEnabled() {
     return mPreferences.getBoolean(PREFS_JS_DEV_MODE_DEBUG_KEY, true);
   }
 
+  @Override
+  public  boolean isUseCachedBundleEnabled() {
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/devsupport/DevInternalSettings.java]: `shouldUseCachedBundle` for consistency with `getShouldUseCachedBundle`

**Comment by mkonicek** [comment]
I like this, it's really annoying we try to load the bundle; this always fails for OSS apps.

Could we specify this to be `true` by default for all OSS apps? For example, here: https://github.com/facebook/react-native/blob/master/local-cli/generator-android/templates/package/MainActivity.java#L24

That way people won't have to set this setting manually and fb apps will still have this set to false.

**Comment by mkonicek** [comment]
Alternatively, can we never read from file if the packager is running? I would expect the check here to do that but apparently it doesn't: https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/ReactInstanceManagerImpl.java#L291

**Comment by qbig** [comment]
Hi @mkonicek . Sorry but may I ask what is "OSS apps" ?

**Comment by qbig** [comment]
@ide @mkonicek the current PR achieved 'never load from cache' by setting in "Dev Setting". But even after this setting, the redbox behaviour is still a bit inconsistent as mentioned here #4739.

The update would fix #4739 as well.

**Comment by qbig** [comment]
@skevy could you update the 'needs-revision' flag ?

**Comment by satya164** [comment]
Looks like this has conflicts now

**Comment by qbig** [comment]
@satya164 resolved :)

**Comment by satya164** [comment]
@qbig OSS -> Open Source Software

**Comment by astreet** [comment]
I think we want to get rid of the cached bundle completely instead (let's wait to make sure Krzysztof didn't add it for some reason I'm not thinking of though).

In that case, I'd ask that we update this PR to remove that logic instead of adding a checkbox.

**Comment by satya164** [comment]
I really like having a cached bundle. That means I can test even if I disconnect the app from the laptop an go out. I don't need to remeber having to install the release version every time.

**Comment by astreet** [comment]
Fair point -- has it bothered you not having it on iOS?

**Comment by satya164** [comment]
@astreet I've only worked on Android so far :)

**Comment by astreet** [comment]
Got ya. Ok, I can see that being a pretty strong usecase, I'll comment on the PR as-is then.

File: ReactAndroid/src/main/java/com/facebook/react/ReactInstanceManagerImpl.java [hunk 3]
```diff
@@ -289,7 +289,7 @@ public void onPackagerStatusFetched(final boolean packagerIsRunning) {
                     new Runnable() {
                       @Override
                       public void run() {
-                        if (packagerIsRunning) {
+                        if (packagerIsRunning || !mDevSupportManager.hasUpToDateJSBundleInCache()) {
```
Comment (astreet) [inline ReactAndroid/src/main/java/com/facebook/react/ReactInstanceManagerImpl.java]: Why this change? We check it above on L278

File: ReactAndroid/src/main/res/devsupport/xml/preferences.xml [hunk 4]
```diff
@@ -19,6 +19,12 @@
         android:summary="At the end of animations, Toasts and logs to logcat debug information about the FPS during that transition. Currently only supported for transitions (animated navigations)."
         android:defaultValue="false"
         />
+    <CheckBoxPreference
+        android:key="use_cached_bundle"
+        android:title="Use Cached Bundle"
+        android:summary="Use prefetched JS bundle stored locally."
+        android:defaultValue="false"
```
Comment (astreet) [inline ReactAndroid/src/main/res/devsupport/xml/preferences.xml]: I would say we should make this default to `true`

File: ReactAndroid/src/main/res/devsupport/xml/preferences.xml [hunk 5]
```diff
@@ -19,6 +19,12 @@
         android:summary="At the end of animations, Toasts and logs to logcat debug information about the FPS during that transition. Currently only supported for transitions (animated navigations)."
         android:defaultValue="false"
         />
+    <CheckBoxPreference
+        android:key="use_cached_bundle"
+        android:title="Use Cached Bundle"
+        android:summary="Use prefetched JS bundle stored locally."
```
Comment (astreet) [inline ReactAndroid/src/main/res/devsupport/xml/preferences.xml]: It's not really prefetching: I'd update this to say something like `Use the last JS bundle downloaded from the packager if the packager cannot be reached`

File: ReactAndroid/src/main/java/com/facebook/react/ReactInstanceManagerImpl.java (hunk 3 above)
Comment (qbig) [inline ReactAndroid/src/main/java/com/facebook/react/ReactInstanceManagerImpl.java]: @astreet  I did this for #4739 so that, when the packager is not running and there is no update-to-date bundle, it would try to load a existent bundle and throw a confusing redbox message( the `else` part)

Comment (qbig) [inline ReactAndroid/src/main/java/com/facebook/react/ReactInstanceManagerImpl.java]: maybe I will just not touch it and fix the redbox message in a different PR

**Comment by qbig** [comment]
@astreet  updated accordingly. Would love to hear your thought on #4739

File: ReactAndroid/src/main/java/com/facebook/react/devsupport/DevInternalSettings.java [hunk 6]
```diff
@@ -72,7 +78,7 @@ public boolean isJSDevModeEnabled() {
   public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
     if (PREFS_FPS_DEBUG_KEY.equals(key) ||
         PREFS_RELOAD_ON_JS_CHANGE_KEY.equals(key) ||
-        PREFS_JS_DEV_MODE_DEBUG_KEY.equals(key)) {
+        PREFS_JS_DEV_MODE_DEBUG_KEY.equals(key) || PREFS_USE_CACHED_BUNDLE.equals(key)) {
```
Comment (astreet) [inline ReactAndroid/src/main/java/com/facebook/react/devsupport/DevInternalSettings.java]: nit: can you move this to a new line like all the other checks?

**Comment by astreet** [comment]
Seems reasonable, see nit inline and I'll merge

**Comment by qbig** [comment]
@astreet fixed

**Comment by astreet** [comment]
Thank you!

**Comment by astreet** [comment]
@facebook-github-bot shipit

**Comment by qbig** [comment]
@astreet Anything I could improve here ?

**Comment by mkonicek** [comment]
There's a compilation error, scroll down in the CircleCI results:
https://circleci.com/gh/facebook/react-native/1287

**Comment by astreet** [comment]
Yay automated testing! Thanks Martin :)

**Comment by mkonicek** [comment]
:trophy:

**Comment by jsierles** [comment]
@astreet Hey, just pinging to see what the status of this PR is. Thanks!

**Comment by mkonicek** [comment]
@qbig Can you rebase please and resolve any failing tests?

**Comment by facebook-github-bot-1** [comment]
@qbig do you have any updates for this pull request? It's been a while so we wanted to check in and see if you've looked at the requested changes.

**Comment by mkonicek** [comment]
Let's close this since there hasn't been a response from the author. If you want to continue working on this please send a new pull request based on top of master.
